### PR TITLE
Integrate checklist component for custom task views

### PR DESCRIPTION
**Pull request by jrbotros**
Submitting for initial code review; user research pending.

Data object is passed into the orchestraChecklist directive; if the data belongs to `taskAssignment.task.data`, it is persisted to the db on save/submit.

### CODE REVIEW

File: orchestra/static/orchestra/common/components/checklist/css/checklist.scss [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+$checkbox-font-size: 22px;
+
+.disable-select {
+  -webkit-user-select: none;
+  -moz-user-select: none;
+  -ms-user-select: none;
+  user-select: none;
+}
+
+@mixin vendor-prefix($name, $argument) {
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/css/checklist.scss]: can this be moved into some global file?

File: orchestra/static/orchestra/common/components/checklist/css/checklist.scss [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+  -webkit-#{$name}: #{$argument};
+  -ms-#{$name}: #{$argument};
+  -moz-#{$name}: #{$argument};
+  -o-#{$name}: #{$argument};
+  #{$name}: #{$argument};
+}
+
+.all-transition {
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/css/checklist.scss]: will this affect all transitions on any aspect of the site?  dangerous?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    return {
+      restrict: 'E',
+      scope: {
+        'data': '=',
+        'author': '=?',
+        'imagePrefix': '=?',
+      },
+      link: function(scope, el, attr) {
+        var sampleItem = {
+          'title': 'This is a sample checklist item. ' +
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js]: From user research it appears we learned that folks might be tempted to use this with coarse granularity (e.g., "Review 1") rather than fine granularity (e.g., "Fix the spelling of `teh`").  Can the example here be biased toward fine granularity so folks can be primed to use it that way?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+                   'Double-click on the text to edit, ' +
+                   'or click to the right to expand.',
+          'comments': [
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:21]: similarly: can we prime folks to view this as a conversation rather than a single explanatory comment?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        scope.data = scope.data || {};
+        scope.data.items = scope.data.items || [sampleItem];
+        scope.data.deletedItems = scope.data.deletedItems || [];
+
+        orchestraService.signals.registerSignal('submit.before', function() {
+          scope.data.items.forEach(function(item) {
+            if (item.comments[item.comments.length - 1].text) {
+              item.comments[item.comments.length - 1].author = scope.author;
+              item.comments[item.comments.length - 1].timestamp = Date.now();
+              item.comments.push({'author': '', 'text': ''});
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js]: ?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        // Initialize checklist data for a fresh task
+        scope.data = scope.data || {};
+        scope.data.items = scope.data.items || [sampleItem];
+        scope.data.deletedItems = scope.data.deletedItems || [];
+
+        orchestraService.signals.registerSignal('submit.before', function() {
+          scope.data.items.forEach(function(item) {
+            if (item.comments[item.comments.length - 1].text) {
+              item.comments[item.comments.length - 1].author = scope.author;
+              item.comments[item.comments.length - 1].timestamp = Date.now();
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:62]: does this store a timezone?  ideally, it would store all times in UTC

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+          'index': 0,
+          'order': 0
+        };
+
+        // Initialize checklist data for a fresh task
+        scope.data = scope.data || {};
+        scope.data.items = scope.data.items || [sampleItem];
+        scope.data.deletedItems = scope.data.deletedItems || [];
+
+        orchestraService.signals.registerSignal('submit.before', function() {
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:56]: comment on what's happening here

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+          scope.data.items.forEach(function(item) {
+            if (item.comments[item.comments.length - 1].text) {
+              item.comments[item.comments.length - 1].author = scope.author;
+              item.comments[item.comments.length - 1].timestamp = Date.now();
+              item.comments.push({'author': '', 'text': ''});
+            }
+          });
+        });
+
+        // Revert commit of latest comments if submit fails
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js]: commit->details

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        orchestraService.signals.registerSignal('submit.error', function() {
+          scope.data.items.forEach(function(item) {
+            if (item.comments[item.comments.length - 1].text) {
+              item.comments.pop();
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:72]: what about removing the author/date?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            }
+          });
+        });
+
+        // Setup drag to reorder
+        scope.preventAction = false;
+        var checklistContainer = el.find('ul.checklist');
+        var drake = dragula([checklistContainer.get(0)], {
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:82]: why is it called drake?

Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:82]: i remember you said https://github.com/bevacqua/angular-dragula was heavyweight---a quick word on why we decided against it?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+                                 'order': scope.data.items.length});
+        }
+
+        // Remove checklist item
+        scope.removeItem = function(item, confirm_needed) {
+          var index = scope.data.items.indexOf(item);
+          if (!confirm_needed || confirm('Are you sure you want to delete this item?')) {
+            scope.preventAction = true;
+            scope.data.items.splice(index, 1);
+            scope.data.deletedItems.push(item);
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:148]: do we display these?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+            }
+            scope.data.items.forEach(function(scopeItem) {scopeItem.readonly = true})
+          }
+          item.readonly = !item.readonly;
+        }
+
+        // Expand checklist item to show comments
+        scope.expandItem = function(item) {
+          item.expanded = !item.expanded;
+          scope.data.items.forEach(function(scopeItem) {scopeItem.readonly = true})
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:174]: what is this?

File: orchestra/static/orchestra/common/components/checklist/js/directives.js [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+        }
+
+        // Return localized comment timestamp
+        scope.getCommentTimestamp = function(comment) {
+          return new Date(comment.timestamp).toLocaleString();
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:179]: does this have a timezone, and is it UTC?

File: orchestra/static/orchestra/common/components/checklist/partials/checklist.html [hunk 14]
```diff
@@ -0,0 +1,21 @@
+<div class="checklist-wrapper">
+    <div ng-if="data.items.length" class="toggle-completed"
+         ng-click="data.hideCompleted = !data.hideCompleted">
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/partials/checklist.html:3]: I can't understand from the code whether the default for `hideCompleted` is true or false.  i think it should be false.

File: orchestra/static/orchestra/common/components/checklist/partials/checklist-item.html [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        <div ng-repeat="comment in item.comments track by $index"
+             class="comment">
+            <div class="comment-meta">
+              <div ng-if="!$last" class="comment-author">{{comment.author}}</div>
+              <div ng-if="!$last" class="comment-timestamp">{{getCommentTimestamp(comment)}}</div>
+            </div>
+            <div class="add-comment" ng-if="$last">Add a comment:</div>
+            <orchestra-quill readonly="!$last"
+                             data="comment.text"
+                             image-prefix="imagePrefix">
```
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/partials/checklist-item.html:51]: should this be `{{imagePrefix}}`?

File: orchestra/static/orchestra/common/components/checklist/css/checklist.scss (hunk 2 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/css/checklist.scss]: Nope, named it `.all-transition` because it just transitions all available CSS attributes on anything that extends the class.

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 6 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:62]: It currently stores the epoch and renders it to local time when it's displayed. Let me know if that sounds fine!

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 8 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js]: This reverts the marking of the latest comments as "readonly"—I'll just be more explicit here.

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 7 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:56]: :white_check_mark:

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 8 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js]: :white_check_mark:

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 9 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:72]: :white_check_mark:

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 10 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:82]: That's the name the author uses to refer to the API instance so I thought I'd be consistent, but happy to change it to whatever.

Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:82]: It wasn't clear to me how the data is persisted in angular-dragula, but it seems like it just stores the HTML contents of the draggable containers. I definitely could be wrong, though. Also, you have to register it in the orchestra app definition rather than only in the steps that use it.

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 11 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:148]: Not right now, let me know if you think we should.

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 12 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:174]: Toggling an expand doesn't remove the focus from the checklist title input; we want to switch the user out of "edit mode" when the user clicks to expand.

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 13 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:179]: Handled by the browser, should be local time: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleTimeString

There are definitely more sophisticated ways of handling this, but we'd need to pick a third-party library—let me know if this is something you're imminently worried about.

File: orchestra/static/orchestra/common/components/checklist/partials/checklist.html (hunk 14 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/partials/checklist.html:3]: It's undefined on init, so falsy, but explicit is better than implicit.

Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/partials/checklist.html:3]: :white_check_mark:

File: orchestra/static/orchestra/common/components/checklist/partials/checklist-item.html (hunk 15 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/partials/checklist-item.html:51]: I guess if we're worried about the checklist scope being able to change the parent scope, we should do that and replace `'imagePrefix': '=?'` with `'imagePrefix': '@?'`.

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 3 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js]: Rewrote the sample item, let me know what you think!

screen shot 2015-10-13 at 4 23 16 pm

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 4 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:21]: See above :white_check_mark:

File: orchestra/static/orchestra/common/components/checklist/partials/checklist-item.html (hunk 15 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/partials/checklist-item.html:51]: I went ahead and made that change in the checklist directive; I'll add the quill fix in with the other changes I have planned.

File: orchestra/static/orchestra/common/components/checklist/css/checklist.scss (hunk 1 above)
Comment (jrbotros) [inline orchestra/static/orchestra/common/components/checklist/css/checklist.scss]: Apparently! Let me submit another PR reorganizing our stylesheets, I think it's time we cleaned things up.

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 6 above)
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:62]: lgtm---it's in UTC when stored, so i'm happy :)

File: orchestra/static/orchestra/common/components/checklist/js/directives.js (hunk 11 above)
Comment (marcua) [inline orchestra/static/orchestra/common/components/checklist/js/directives.js:148]: ok for now:P

**Comment by marcua** [comment]
lgtm!
